runtime/onert/backend/cpu/ops/MaxLayer.cc

```diff
@@ -38,12 +38,16 @@ void MaxLayer::maxFloat32()
 
 void MaxLayer::maxQuant8()
 {
-  // TODO Check whether cker for quant8 max produces correct results
-  // nnfw::cker::Max<uint8_t>(
-  //     getTensorShape(_lhs), reinterpret_cast<const uint8_t*>(_lhs->buffer()),
-  //     getTensorShape(_rhs), reinterpret_cast<const uint8_t*>(_rhs->buffer()),
-  //     getTensorShape(_output), reinterpret_cast<uint8_t*>(_output->buffer()));
-
+  if (_lhs->data_scale() == _rhs->data_scale() && _lhs->data_scale() == _output->data_scale())
```

Comment:
I think we should make `MaxLayer` supports qasymm8 if `scale` and `offset` are different. Are you going to do it?

Reply:
I agree.
actually we already have testcase with diffrent scale and offset. (**maximum.mod.py** source file)

but according to following [document](https://www.tensorflow.org/lite/performance/quantization_spec).
tensorflow seems to support some cases with the same scale and offset.

But this document looks old.
If I'm wrong, please let me know to fix this PR.

Reply:
Tensorflow lite kernel for `Maximum` operation don't support qasymm8 for different `scale` and `offset`. So we may not need to support different `scale` and `offset` for tflite (and circle) model inference now.

If we need to support different `scale` and `offset` to support NNAPI specification and pass nnapi generation test, we should port android nn runtime's kernel, not tensorflow lite kernel. But it is not our requirement.

Reply:
@hasw7569 

> but according to following [document](https://www.tensorflow.org/lite/performance/quantization_spec).
tensorflow seems to support some cases with the same scale and offset.
`
MAXIMUM
  Input 0:
    data_type  : int8
    range      : [-128, 127]
    granularity: per-tensor
  Output 0:
    data_type  : int8
    range      : [-128, 127]
    granularity: per-tensor
  restriction: Input and outputs must all have same scale/zero_point
`
Tensorflow lite kernel for Maximum operation (2.3-rc0 version) supports only 
It seems to describes about int8 quantization . But you implemented to support `QUANT_UINT8_ASYMM`. Both are different quantization.

Until NNAPI V1.2(Android API level 29), Maximum operation supports `TENSOR_QUANT8_ASYMM`, but not `TENSOR_QUANT8_ASYMM_SIGNED`. Our runtime supports until nnapi v1.2 now.
Tensorflow lite kernel for Maximum operation (2.3-rc0 version) supports signed quant8(`TENSOR_QUANT8_ASYMM_SIGNED`).
So, I think we implement these two quantizations separately.
1. Supporting `TENSOR_QUANT8_ASYMM` for the nnapi unitest to work
2. Supporting `TENSOR_QUANT8_ASYMM_SIGNED` for a test tflite model to works

If your goal was to supporting a tflite model, you should implements it as second way.

Reply:
@ragmini @hasw7569 

As @ragmini commented, document https://www.tensorflow.org/lite/performance/quantization_spec describes in8 quantization. So to check tensorflow lite's quantization spec, we should check kernel implementation.

But as I commented above, tensorflow lite (2.3-rc0)'s `uint8 qasymm8` kernel also support `maximum` kernel for same `scale` and `offset`. And other operations `uint8 qasymm8` kernel also same spec with `int8`. 

After we defined `nnfw api`, `NNAPI` is not official API for our runtime, so `NNAPI` frontend is optional frontend to use tensorflow lite and test code (nnapi generation tests).

Reply:
> But as I commented above, tensorflow lite (2.3-rc0)'s `uint8 qasymm8` kernel also support `maximum` kernel for same scale and offset.

It is strange. Why does the kernel support only the type for same `scale` and `offset`. Tensorflow lite (2.3-rc)'s nnapi delegate converts `TfLiteQuantizationParams` to `ANEURALNETWORKS_TENSOR_QUANT8_ASYMM` as it is? How  does tensorflow lite makes it has inputs and output of the operation as the same `scale` and `offset`? Does `TOCO` or TFLC make a model with `maximum` operation like that? I don't know it.
Anyway, I think we cannot sure `uint qasymm8` type of `maximum` operation until we see a model that includes the operation in the type. 
@hasw7569 Have you seen any model with `maximum` operation in `uint qasymm8`?


>  And other operations `uint8 qasymm8` kernel also same spec with `int8`.

@hseok-oh 
I disagree with that other operations `uint8 qasymm8` kernel also same spec with `int8`. `uint qasymm8` and `int8` are different spec. All `int8` can use the kernels for `uint qasymm8`, but `uint qasymm8` cannot use the kernels for `int8`. Only when `zero_point` is 128, it is possible.


> After we defined nnfw api, NNAPI is not official API for our runtime, so NNAPI frontend is optional frontend to use tensorflow lite and test code (nnapi generation tests).

I agree with you.

Reply:
>How does tensorflow lite makes it has inputs and output of the operation as the same scale and offset? Does TOCO or TFLC make a model with maximum operation like that? I don't know it.

I don't know. But tensorflow lite kernel assume that input & output have same `scale` and `offset`, so maybe `TFLC` make them same. 

https://github.com/tensorflow/tensorflow/blob/99fea8da0d98fb271b60b58cfa5755f2bd430079/tensorflow/lite/kernels/maximum_minimum.cc#L79-L84

```cpp
struct MaximumOp {
  template <typename data_type>
  static data_type op(data_type el1, data_type el2) {
    return el1 > el2 ? el1 : el2;
  }
};
```

https://github.com/tensorflow/tensorflow/blob/99fea8da0d98fb271b60b58cfa5755f2bd430079/tensorflow/lite/kernels/maximum_minimum.cc#L93-L104
```cpp
template <KernelType kernel_type, typename data_type, typename op_type>
void TFLiteOperation(TfLiteContext* context, TfLiteNode* node,
                     const OpContext& op_context) {
  reference_ops::MaximumMinimumBroadcastSlow(
      GetTensorShape(op_context.input1),
      GetTensorData<data_type>(op_context.input1),
      GetTensorShape(op_context.input2),
      GetTensorData<data_type>(op_context.input2),
      GetTensorShape(op_context.output),
      GetTensorData<data_type>(op_context.output),
      op_type::template op<data_type>);
}
```

https://github.com/tensorflow/tensorflow/blob/99fea8da0d98fb271b60b58cfa5755f2bd430079/tensorflow/lite/kernels/internal/reference/maximum_minimum.h#L52-L57
```cpp
    auto maxmin_func = [&](int indexes[N]) {
      output_data[SubscriptToIndex(output_desc, indexes)] =
          op(input1_data[SubscriptToIndex(desc1, indexes)],
             input2_data[SubscriptToIndex(desc2, indexes)]);
    };
    NDOpsHelper<N>(output_desc, maxmin_func);
```

Reply:
>I disagree with that other operations uint8 qasymm8 kernel also same spec with int8

`uint8 qasymm8` kernel and `int8` kernel don't share code. `uint8 qasymm8` kernel is just legacy code and `int8` kernel is different code. But they have same spec in document for `int8` when I checked each operations. If `int8` allow different `scale` and `offset`, `uint8 qasymm8` also allow. If `int8` support only same `scale` and `offset`, `uint8 qasymm8` also support same `scale` and `offset.

Reply:
@hseok-oh 

> I don't know. But tensorflow lite kernel assume that input & output have same scale and offset, so maybe `TFLC` make them same.

I hope so.